design-proposals/sdn-ipam-secondary-nets.md

```diff
@@ -0,0 +1,490 @@
+  IPAM CNI->>apiserver: IPAMClaim.UpdateStatus(status.ips = ips)
+  apiserver-->>IPAM CNI: OK
+  end
+
+  deactivate CNI
+  apiserver->>CNI: reconcileIPAMClaims()
+  loop for ipamClaim := range pod.ipamClaims
+  CNI->>CNI: configureIPs(pod, ipamClaim.ips)
+  end
+```
+
+**NOTES**:
+- When the feature gate is not enabled, KubeVirt will **not** request
+  persistent IP allocations from the CNI.
+- OVN-Kubernetes will **not** check for consistency errors related to the
+  `IPAMClaim`s used in two different pods simultaneously.
+
+#### Stopping a Virtual Machine
+```mermaid
+sequenceDiagram
+actor user
+
+user ->> KubeVirt: stopVM(vm=vmA)
+
+KubeVirt ->> Kubernetes: deletePod(owner=vmA)
+Kubernetes -->> KubeVirt: OK
+
+KubeVirt -->> user: OK
+
+note over OVN-K: notices a pod was deleted
+OVN-K ->> OVN-K: ips = teardownOVNEntities()
+```
+
+OVN-Kubernetes will **not** delete the `IPAMClaim` CRs for the deleted VM pod.
+The IP addresses allocated to the VM will **not** be released to the IP pool
+when there's a corresponding `IPAMClaim` for the attachment being deleted.
+
+#### Starting a (previously stopped) Virtual Machine
+This flow is - from a CNI perspective - quite similar to the
+[Creating a VM flow](#creating-a-virtual-machine):
+1. the workload controller (KubeVirt) templates the pod, featuring the required
+   `IPAMClaim` references in each network selection element.
+2. OVN-Kubernetes will read the network selection element; if it sees an
+   `IPAMClaim` reference, it will attempt to read it from the datastore.
+3. if the read `IPAMClaim` features IP addresses in its status, OVN-Kubernetes
+   will configure the interface with those IPs. If the status doesn't feature
+   IP addresses, the CNI plugin will retry later, with exponential backoff.
+
+As when creating a new VM, an error is thrown when the IPAMClaim belonged to an
+old VM with the same name. Eventually the garbage collector will remove this old
+IPAMClaim, and in the next reconcile loop the controller will create a new
+IPAMClaim.
+
+```mermaid
+sequenceDiagram
+
+  actor user
+  participant KubeVirt
+  participant Kubernetes
+  participant CNI
+
+  user->>KubeVirt: startVM(vmName)
+
+  note over KubeVirt: podName := "launcher-<vm name>"
+  KubeVirt->>Kubernetes: createPod(name=podName, ipam-claims=[claimName])
+
+  Kubernetes->>CNI: CNI ADD
+  note over CNI: this CNIs has an IPAMClaims informer
+
+  CNI->>CNI: ipamClaim := readIPAMClaim(claimName)
+  CNI->>CNI: configIface(ipamClaim.Status.IPs)
+
+  CNI-->>Kubernetes: OK
+
+  Kubernetes-->>KubeVirt: OK
+
+
+  KubeVirt-->>user: OK
+```
+
+When creating the pod, the OVN-Kubernetes IPAM module finds existing
+`IPAMClaim`s for the workload. It will thus use those already reserved
+allocations, instead of generating brand new allocations for the pod where the
+encapsulating object will run. The migration scenario is similar.
+
+#### Removing a Virtual Machine
+This flow is - from a CNI perspective - quite similar to the
```

Comment:
assuming we want to support that (I pretty much don't care about it, but can go either way), we just set the VMI's data as owner reference. 

The exact same thing will happen, but the owner is a VMI instead of a VM.

Reply:
I still cannot find foreground and background deletion explained in the doc.
IIRC, foreground was a big issue when thinking about the design (what happens if the IPAMClaim is removed before the VM). I believe it is important to describe the problem and how the current design solves it in the doc. I would expect to see an explanation how do we prevent the IPAMClaim from being removed before the VM, VMI and virt-launcher pod.

Reply:
Done. 

Let me know if this is good enough for what you have in mind.

Reply:
Thanks, still missing - in case of background deletion  - how do you make sure the IPACMClaim is not removed before the VMI/virt-launcher pod.

Reply:
The VMI has a finalizer that is removed by kubevirt only when the launcher pod is gone. 

Right @oshoval ?

Reply:
yep, and VM <-> VMI are also protected with mutual finalizers
so only when the VMI is deleted (or pod in case of standalone) we will remove the finalizer
as done in the current logic

Reply:
Can you please add this explanation to the doc?

Reply:
Done.